### ISSUE TITLE
Use s2i for containerized image builds

### ISSUE THREAD

**Issue by andrewazores** [cryostatio/cryostat#1]
Build tooling should be switched from `jib` to `s2i`.

**Comment by andrewazores** [comment]
Also needs to be applied to https://github.com/rh-jmc-team/container-jfr-reports-generator .

**Comment by tabjy** [comment]
In order to use `s2i`, a specialized builder image seems to be needed. Such an image must satisfy the following requirements:

- openjdk11
- maven
- gradle (not if using gradle wrapper)
- git
- npm

I've created a very simple demo image, alone with some s2i scripts, that does the job. See [tabjy/container-jfr-builder](https://github.com/tabjy/container-jfr-builder).

Q: Why are maven, git and npm necessary?
A: Needed for cloning and building JMC, container-jfr-core, and container-jfr-web, which are dependencies of this application.

Q: Why not installing build tools in s2i scripts just before building, rather than creating a specialized image.
A: That wouldn't work in Openshift. By default, Openshift doesn't allow processes in containers to be executed as root.

**Comment by andrewazores** [comment]
Looks like a great first pass at this, thanks!

The Dockerfile for the base image seems fine, although I think it might be better to explicitly set the alternatives paths rather than choosing the 2nd option.

My s2i build is currently failing with this:

```
[ERROR] Failed to execute goal org.reficio:p2-maven-plugin:1.3.0:site (default-cli) on project external-dependencies: Execution default-cli of goal org.reficio:p2-maven-plugin:1.3.0:site failed: org.apache.maven.plugin.MojoExecutionException: Unable to execute mojo: P2 publisher return code was 13 -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
Build failed
```

But this looks like it might just be a temporary outage issue? Not sure since I can't see the log file.

Otherwise the end image build also looks fine. It's worth considering shallow cloning and specifying a commit hash or release tag for the upstream JMC and container-jfr-core too, ex. `git clone --branch v0.2.4 --depth 1 https://rh-jmc-team/container-jfr-core`. No need for all the version history for this kind of use case.

I'd really like to see the TODOs you've listed addressed, especially caching to speed up the build process.

**Comment by tabjy** [comment]
The `mvn p2:site` failure is a known issue. See tabjy/container-jfr-builder#1 for details. I have found a solution to this, and I'm currently working on a fix.

Shallow cloning will be added to the list of todos, and I'll continue working on those todos.

**Comment by tabjy** [comment]
[tabjy/container-jfr-builder](https://github.com/tabjy/container-jfr-builder) is updated with following changes: 

- Fixed tabjy/container-jfr-builder#1
- Cache for maven, gradle, npm (enabled with incremental build strategy)
- Explicitly set alternatives paths for Java executables
- Support shallow clone and cloning the specific branch

Please give it a try. Thanks!

**Comment by andrewazores** [comment]
The incremental builds are definitely much faster than otherwise, but it's still quite the slow process overall. It seems that the Maven dependencies are cached, but the actual built artefacts perhaps aren't?

**Comment by tabjy** [comment]
From commit [042dc47](https://github.com/tabjy/container-jfr-builder/commit/042dc476a598e1e9665ee2cd99f057a9d580b77e), `assemble` script checks for existing containerjfr-core and JMC artefacts, and only build them if not found. It should be much faster now.

**Comment by andrewazores** [comment]
Much better. I'm curious to try this out with the s2i build strategy in an Openshift deployment - the actual image build, even incremental, is still significantly slower than a jib build, but the s2i immediate rebuild and rollout might be an even better replacement during development cycles.

The last thing I think needs to be addressed before we can use this builder for "official" builds for publication is that the JMC revision used is always the latest upstream master at the time of the image build. That's acceptable for a development build but not how we should be building images that end up actually in use somewhere, since if upstream is updated then we can't really reproduce the image build. Maybe you can add a check for an environment variable, similar to what you did for the container-jfr-core tagged versions, and handle the JMC clone depending on how that variable is set. If it's unset/empty, do a shallow clone of the latest commit. If it is set, then clone the whole repository history (unfortunately) and check out the specified revision. Then in the future we can lobby for the JMC mirror to have tags for each minor version and clean up that clone handling.

**Comment by tabjy** [comment]
From what I understand, `jib` builds the application on the local host system and is, therefore, using the local ~/.m2, ~/.npm and ./.gradle directly, then injecting built artifacts into runtime image. `s2i` incremental build has to go through the process of pulling the last runtime image, collecting artifacts into a tarball, inflating and injecting into the build-time image, and then finally start building.

In other words, `s2i` is bound to be slower than `jib`, and there is nothing we can do about it really.

It also came across my mind that since we're cloning from GitHub anyway, we might as well just download the zip archives provided by the platform. Is this viable?

**Comment by andrewazores** [comment]
> In other words, s2i is bound to be slower than jib, and there is nothing we can do about it really.

Yes, agreed, but I don't think the margin necessarily needs to be that large with carefully selected caching. Inflating/deflating an uncompressed tarball isn't that expensive of an operation, so it should be possible to get the incremental build fairly quick.

> It also came across my mind that since we're cloning from GitHub anyway, we might as well just download the zip archives provided by the platform. The zip archive is always shallow and one can specify any revision in history. Is this viable?

That's a good observation, I didn't realize GitHub provides these zip archives by revision even without tags. May as well take advantage then.

**Comment by tabjy** [comment]
Revision [f4a96b0](https://github.com/tabjy/container-jfr-builder/commit/f4a96b0c8a533e96e949e60882aef7482ceeb5ff) nows uses zip archives instead of git. Dependency on `git` is removed.

Also, I'm not sure what you mean by "carefully selected caching"? Since the first build starts with an empty `.m2` repo, weren't all contents in `.m2` after the build is required? The same goes for `.npm`.

**Comment by andrewazores** [comment]
Sure, the contents of `.m2` and `.npm` should all be required for the next incremental build, but that doesn't imply that all of what is required for the next rebuild are contained in those two directories. The build of `container-jfr-core` installs its resultant artefacts into `.m2`, but the build of `container-jfr-web` does not end up in `.npm` - so when you attempt a rebuild, that component has to be entirely rebuilt even if the sources have not changed, because the results of the previous build were "lost" (yes, they are preserved in the final tarball, but the next time `npm run build` occurs, it will see an empty `dist` subdirectory).
